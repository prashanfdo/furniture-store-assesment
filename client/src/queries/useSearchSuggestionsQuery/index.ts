import { useQuery } from "@apollo/client";
import SEARCH_SUGGESTIONS_QUERY from "./query";
import { SearchSuggestionsQueryData, SearchSuggestionsQueryVars } from "./useSearchSuggestionsQuery.types";
import { DEFAULT_LOCALE } from "helpers/constants";

export default function useSearchSuggestionsQuery(searchText: string) {
  return useQuery<SearchSuggestionsQueryData, SearchSuggestionsQueryVars>(SEARCH_SUGGESTIONS_QUERY, {
    variables: { prefix: searchText, locale: DEFAULT_LOCALE },
  });
}
