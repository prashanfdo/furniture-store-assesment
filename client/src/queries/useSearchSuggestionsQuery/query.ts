import { gql } from "@apollo/client";

const SEARCH_SUGGESTIONS_QUERY = gql`
  query GetSearchSuggestions($prefix: String!, $locale: Locale!) {
    searchSuggestions: autoSuggestion(prefix: $prefix, locale: $locale) {
      name
      results
      image(format: WEBP, maxWidth: 32, maxHeight: 32)
    }
  }
`;

export default SEARCH_SUGGESTIONS_QUERY;
