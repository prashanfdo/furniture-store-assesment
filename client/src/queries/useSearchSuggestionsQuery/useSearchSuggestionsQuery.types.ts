export type SearchSuggestionsQueryData = {
  searchSuggestions: {
    name: string;
    results: number;
    image: string;
  }[];
};

export type SearchSuggestionsQueryVars = {
  prefix: string;
  locale: string;
};
