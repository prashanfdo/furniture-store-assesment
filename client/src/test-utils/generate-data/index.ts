export * from "./article";
export * from "./category";
export * from "./search";
export * from "./gql";
