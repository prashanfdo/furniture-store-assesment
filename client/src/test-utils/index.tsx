import { render, RenderOptions, renderHook, RenderHookOptions, RenderHookResult } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import user from "@testing-library/user-event";
import { ProvidersForTests } from "providers";

type GraphQlMocks = React.ComponentProps<typeof ProvidersForTests>["mocks"];

type WrapperProps = React.ComponentProps<typeof ProvidersForTests> & {
  children: React.ReactNode;
};
const Wrapper: FC<WrapperProps> = ({ children, mocks, initialRoute }: WrapperProps) => {
  return (
    <ProvidersForTests mocks={mocks} initialRoute={initialRoute}>
      {children}
    </ProvidersForTests>
  );
};
const wrapper =
  (mocks: GraphQlMocks, initialRoute?: string) =>
  ({ children }: { children: React.ReactNode }) =>
    (
      <Wrapper mocks={mocks} initialRoute={initialRoute}>
        {children}
      </Wrapper>
    );

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper"> & { gqlMocks: GraphQlMocks; initialRoute: string },
) => {
  const view = render(ui, { wrapper: wrapper(options?.gqlMocks, options?.initialRoute), ...options });
  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};

export const customRenderHook = <T, R>(
  callback: (props: T) => R,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { gqlMocks?: GraphQlMocks; initialRoute?: string },
): RenderHookResult<R, T> => {
  return renderHook(callback, { ...options, wrapper: wrapper(options?.gqlMocks, options?.initialRoute) });
};

export { faker } from "@faker-js/faker";
export * from "@testing-library/react";
export * from "./generate-data";
export { customRender as render };
export { customRenderHook as renderHook };
export { user };
