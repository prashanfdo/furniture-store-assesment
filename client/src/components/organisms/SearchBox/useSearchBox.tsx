import useSearchSuggestionsQuery from "queries/useSearchSuggestionsQuery";
import { useState } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

// TODO: add comments
const useSearchBox = () => {
  const navigate = useNavigate();
  const [showSuggestions, setShowSuggestions] = useState<boolean>(false);

  const [searchParams] = useSearchParams();
  const initialSearchText = searchParams.get("q") || "";
  const [searchText, setSearchText] = useState<string>("");

  const handleSearchSubmit = (searchText: string) => {
    setShowSuggestions(false);
    if (!!searchText?.trim()) {
      navigate(`/search?q=${encodeURIComponent(searchText?.trim())}`);
    } else {
      navigate(`/`);
    }
  };

  const handleSearchChange = (newText: string) => {
    if (newText !== searchText) {
      setShowSuggestions(true);
    }
    setSearchText(newText);
  };

  // TODO: handle loading and error states
  const { data, loading, error } = useSearchSuggestionsQuery(searchText);

  const searchSuggestions = data?.searchSuggestions || [];

  return {
    loading,
    error,
    initialSearchText,
    handleSearchChange,
    handleSearchSubmit,
    searchSuggestions,
    showSuggestions,
    setShowSuggestions,
  };
};

export default useSearchBox;
