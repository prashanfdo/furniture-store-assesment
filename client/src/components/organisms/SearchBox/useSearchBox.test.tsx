import { createGqlResult, renderHook } from "test-utils";
import useSearchBox from "./useSearchBox";
import useSearchSuggestionsQuery from "queries/useSearchSuggestionsQuery";
import { useSearchParams } from "react-router-dom";
import {
  SearchSuggestionsQueryData,
  SearchSuggestionsQueryVars,
} from "queries/useSearchSuggestionsQuery/useSearchSuggestionsQuery.types";

jest.mock("queries/useSearchSuggestionsQuery");
const mockUseSearchSuggestionsQuery = useSearchSuggestionsQuery as jest.MockedFunction<
  typeof useSearchSuggestionsQuery
>;

jest.mock("react-router-dom");
const mockUseSearchParams = useSearchParams as jest.MockedFunction<typeof useSearchParams>;

describe("useSearchBox", () => {
  beforeEach(() => {
    mockUseSearchSuggestionsQuery.mockReset();
  });

  it("should render", async () => {
    mockUseSearchSuggestionsQuery.mockReturnValue(
      createGqlResult<SearchSuggestionsQueryData, SearchSuggestionsQueryVars>({
        loading: true,
      }),
    );
    const { result } = renderHook(() => useSearchBox());
    expect(result.current).toMatchSnapshot();
  });

  it("should return initial search text from current url", () => {
    mockUseSearchSuggestionsQuery.mockReturnValue(
      createGqlResult<SearchSuggestionsQueryData, SearchSuggestionsQueryVars>({
        loading: true,
      }),
    );
    mockUseSearchParams.mockReturnValue([new URLSearchParams("searchText=Test"), jest.fn()]);
    const { result } = renderHook(() => useSearchBox());
    expect(result.current).toMatchSnapshot();
    // TODO: fix this test
  });
});

export default useSearchBox;
