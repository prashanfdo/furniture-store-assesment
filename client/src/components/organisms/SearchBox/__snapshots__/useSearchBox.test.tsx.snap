// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSearchBox should render 1`] = `null`;

exports[`useSearchBox should return initial search text from current url 1`] = `null`;
